--- Day 7: No Space Left On Device ---

You can hear birds chirping and raindrops hitting leaves as the expedition proceeds. Occasionally, you can even hear much louder sounds in the distance; how big do the animals get out here, anyway?

The device the Elves gave you has problems with more than just its communication system. You try to run a system update:

    $ system-update --please --pretty-please-with-sugar-on-top
    Error: No space left on device
    Perhaps you can delete some files to make space for the update?

You browse around the filesystem to assess the situation and save the resulting terminal output (your puzzle input). For example:

    $ cd /
    $ ls
    dir a
    14848514 b.txt
    8504156 c.dat
    dir d
    $ cd a
    $ ls
    dir e
    29116 f
    2557 g
    62596 h.lst
    $ cd e
    $ ls
    584 i
    $ cd ..
    $ cd ..
    $ cd d
    $ ls
    4060174 j
    8033020 d.log
    5626152 d.ext
    7214296 k

The filesystem consists of a tree of files (plain data) and directories (which can contain other directories or files). The outermost directory is called /. You can navigate around the filesystem, moving into or out of directories and listing the contents of the directory you're currently in.

Within the terminal output, lines that begin with $ are commands you executed, very much like some modern computers:

    cd means change directory. This changes which directory is the current directory, but the specific result depends on the argument:
    cd x moves in one level: it looks in the current directory for the directory named x and makes it the current directory.
    cd .. moves out one level: it finds the directory that contains the current directory, then makes that directory the current directory.
    cd / switches the current directory to the outermost directory, /.
    ls means list. It prints out all of the files and directories immediately contained by the current directory:
    123 abc means that the current directory contains a file named abc with size 123.
    dir xyz means that the current directory contains a directory named xyz.

Given the commands and output in the example above, you can determine that the filesystem looks visually like this:

- / (dir)
  - a (dir)
    - e (dir)
      - i (file, size=584)
    - f (file, size=29116)
    - g (file, size=2557)
    - h.lst (file, size=62596)
  - b.txt (file, size=14848514)
  - c.dat (file, size=8504156)
  - d (dir)
    - j (file, size=4060174)
    - d.log (file, size=8033020)
    - d.ext (file, size=5626152)
    - k (file, size=7214296)

Here, there are four directories: / (the outermost directory), a and d (which are in /), and e (which is in a). These directories also contain files of various sizes.

Since the disk is full, your first step should probably be to find directories that are good candidates for deletion. To do this, you need to determine the total size of each directory. The total size of a directory is the sum of the sizes of the files it contains, directly or indirectly. (Directories themselves do not count as having any intrinsic size.)

The total sizes of the directories above can be found as follows:

The total size of directory e is 584 because it contains a single file i of size 584 and no other directories.
The directory a has total size 94853 because it contains files f (size 29116), g (size 2557), and h.lst (size 62596), plus file i indirectly (a contains e which contains i).
Directory d has total size 24933642.
As the outermost directory, / contains every file. Its total size is 48381165, the sum of the size of every file.
To begin, find all of the directories with a total size of at most 100000, then calculate the sum of their total sizes. In the example above, these directories are a and e; the sum of their total sizes is 95437 (94853 + 584). (As in this example, this process can count files more than once!)

Find all of the directories with a total size of at most 100000. What is the sum of the total sizes of those directories?

In [118]:
import re

In [162]:
with open("input.txt", "r") as f:
    lines = f.read().splitlines()

In [163]:
# linked lists
class Node():
    def __init__(self, name, parent=None, size=None):
        self.name = name
        self.parent = parent
        self.size = size
        self.children = dict()

    def __repr__(self):
        return f"{self.name}"

    def get_child(self, child_name):
        if child_name in self.children:
            return self.children[child_name]
        else:
            return None

    def get_children(self):
        return list(self.children.values())

    def add_child(self, name, size=None):
        if self.get_child(name):
            print("Child already exists")
            return self.get_child(name)
        else:
            child = Node(name, self, size)
            self.children[name] = child
            return child


    def total_size(self):
        if self.size:
            return self.size
        else:
            return sum([child.total_size() for child in self.children.values()])

In [164]:
root = Node("/")
current = root
dirs = []

for line in lines[1:]:
    if line == "$ ls":
        continue

    elif re.search("(^\d)|(^dir)", line):
        size, name = line.split()
        if size != "dir":
            size = int(size)
            current.add_child(name, size)
        else:
            child = current.add_child(name)
            dirs.append(child)
        # print(f"Adding '{name}' to {current}")

    elif line.startswith("$ cd"):
        dir = line.split()[-1]
        if dir == "..":
            current = current.parent
            # print(f"Going up to {current}")
        else:
            current = current.get_child(dir)
            # print(f"Going down to {current}")

    else:
        print("Missing case:", line)

In [170]:
small_dirs = [dir for dir in dirs if dir.total_size() <= 100000]

sum([dir.total_size() for dir in small_dirs])

1297683

- 25, wrong anser. I counted the number of directories with size <= 100000, but that was not the question. I had to sum the sizes of those directories.
- 1297683 is the right answer.

--- Part Two ---

Now, you're ready to choose a directory to delete.

The total disk space available to the filesystem is 70000000. To run the update, you need unused space of at least 30000000. You need to find a directory you can delete that will free up enough space to run the update.

In the example above, the total size of the outermost directory (and thus the total amount of used space) is 48381165; this means that the size of the unused space must currently be 21618835, which isn't quite the 30000000 required by the update. Therefore, the update still requires a directory with total size of at least 8381165 to be deleted before it can run.

To achieve this, you have the following options:

    Delete directory e, which would increase unused space by 584.
    Delete directory a, which would increase unused space by 94853.
    Delete directory d, which would increase unused space by 24933642.
    Delete directory /, which would increase unused space by 48381165.
    Directories e and a are both too small; deleting them would not free up enough space. However, directories d and / are both big enough! Between these, choose the smallest: d, increasing unused space by 24933642.

Find the smallest directory that, if deleted, would free up enough space on the filesystem to run the update. What is the total size of that directory?

In [169]:
total_space = 70000000
min_available = 30000000
used_space = root.total_size()
available_space = total_space - used_space
space_needed = min_available - available_space
space_needed

4804833

In [179]:
large_enough_dirs = [f"{dir.total_size()} {dir}" for dir in dirs if dir.total_size() >= space_needed]
large_enough_dirs.sort()
large_enough_dirs

['11152310 hmwqgjnl',
 '11351037 zwlpm',
 '43189731 dscbfp',
 '5756764 dscbfp',
 '6600089 wqc',
 '7638482 mmhtpz',
 '8050397 wrb',
 '9427614 qdztzhl',
 '9735449 rgvvz']

5756764 was the right answer. I could have done this in a more efficient way, but I was too lazy.